extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/IndexerWorkerContext.java

```diff
@@ -250,8 +242,8 @@ public FrameContext frameContext(QueryDefinition queryDef, int stageNumber)
         indexIO,
         dataSegmentProvider,
         WorkerMemoryParameters.compute(
-            Runtime.getRuntime().maxMemory(),
-            numWorkersInJvm,
+            computeAvailableHeapMemory(),
```

Comment:
It does need to be done on each worker, because the workers and the controller might have different amounts of memory available to them. (Imagine a job running through a rolling update that alters the amount of available memory.)

Good point that we should avoid needless recomputation of this. I updated things to compute it a single time per worker, as part of `IndexerWorkerContext#createProductionInstance`.

Reply:
I meant doing only the lookup sizing calculation on the controller but I see why it can be misleading :). 
Thanks for the fixes.